crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs

```diff
@@ -913,6 +941,93 @@ impl AcirContext {
         let lhs_less_than_rhs = self.more_than_eq_var(rhs, lhs, bit_size, predicate)?;
         self.maybe_eq_predicate(lhs_less_than_rhs, predicate)
     }
+
+    /// Returns a Variable that is constrained to be the result of reading
+    /// from the memory `block_id` at the given `index`.
+    pub(crate) fn read_from_memory(&mut self, block_id: BlockId, index: &AcirVar) -> AcirVar {
+        // Fetch the witness corresponding to the index
+        //
+        // TODO: We can have a helper method which converts AcirVar to Witness
+        let index_var_data = self.vars.get(index).expect("ICE: undeclared AcirVar");
+        let index_witness = self.acir_ir.get_or_create_witness(&index_var_data.to_expression());
+
+        // Create a Variable to hold the result of the read and extract the corresponding Witness
+        let value_read_var = self.add_variable();
+        let value_read_var_data = self.vars.get(&value_read_var).expect("ICE: undeclared AcirVar");
+        let value_read_witness = value_read_var_data.to_expression().into_owned();
+
+        // Add the memory read operation to the list of opcodes
+        let read_op = Expression::zero();
+        let op =
+            MemOp { operation: read_op, index: index_witness.into(), value: value_read_witness };
+        self.acir_ir.opcodes.push(Opcode::MemoryOp { block_id, op });
+
+        value_read_var
+    }
+
+    /// Constrains the Variable `value` to be the new value located at `index` in the memory `block_id`.
+    pub(crate) fn write_to_memory(&mut self, block_id: BlockId, index: &AcirVar, value: &AcirVar) {
+        // Fetch the witness corresponding to the index
+        //
+        // See read_memory for TODO on creating a helper
+        let index_var_data = self.vars.get(index).expect("ICE: undeclared AcirVar");
+        let index_witness = self.acir_ir.get_or_create_witness(&index_var_data.to_expression());
+
+        // Fetch the witness corresponding to the value to be written
+        let value_write_var_data = self.vars.get(value).expect("ICE: undeclared AcirVar");
+        let value_write_witness =
+            self.acir_ir.get_or_create_witness(&value_write_var_data.to_expression());
+
+        // Add the memory write operation to the list of opcodes
+        let write_op = Expression::one();
+        let op = MemOp {
+            operation: write_op,
+            index: index_witness.into(),
+            value: value_write_witness.into(),
+        };
+        self.acir_ir.opcodes.push(Opcode::MemoryOp { block_id, op });
+    }
+
+    /// Initializes an array in memory with the given values `optional_values`.
+    /// If `optional_values` is empty, then the array is initialized with zeros.
+    ///
+    /// TODO: Can we make `optional_values` Option<&[AcirValue]> ?
```

Comment:
We should not be able to initialize an array with no elements, so I've tentatively made this an Option.

The only place where we use None, is when we are in array_set and we are creating a copy of the array by first initalizing it with zero values and then copying the old array into it. Perhaps we should just remove the optional and have the callsite pass in the constant zero value, since it is only done in one place